File: Hashmap/p383_ransom_note.py
class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        hashmap= defaultdict(int)

        for c in magazine:
            hashmap[c]+=1
        
        print(hashmap)

        for c in ransomNote:
            if c not in hashmap or hashmap[c]==0:
                return False
            
            hashmap[c]-=1
        
        return True

#time complexity O(n+m)
#space complexity O(m)

File: Hashmap/p383_ransom_note.cpp
class Solution {
public:
  bool canConstruct(string ransomNote, string magazine) {
    unordered_map<char, int> noteMap;
    unordered_map<char, int> magMap;

    for (int i = 0; i < ransomNote.size(); i++) {
      noteMap[ransomNote[i]]++;
    }

    for (int i = 0; i < magazine.size(); i++) {
      magMap[magazine[i]]++;
    }

    for (const auto &[key, val] : noteMap) {
      if (magMap[key] < val) {
        return false;
      }
    }

    return true;
    ;
  }
};

// time complexity O(n+m)
// space complexity O(n+m)